### ISSUE TITLE
ValueError: format spec u'x' not recognised with register_type custom type definitions (since 1.2.6)

### ISSUE THREAD

**Issue by belugame** [behave/behave#633]
My code base runs fine on 1.2.5 but when using 1.2.6 my behave tests exit with an error as my custom type definitions apparently are not found anymore.

Has something changed in the way I need to register them? I cannot see something related in the docs or list of changes.

My custom type in utils/features/types.py:
```
from IPy import IP
def parse_ip(value):                                                                                                          
      return IP(value)   

register_type(IP=parse_ip)
```

behave test that triggers error in order/features/steps/ip_address.py:

`@given(u'IP {ip_address:IP} is not a internal IP address')
`

The error:

```
  File "order/features/steps/ip_address.py", line 7, in <module>
    @given(u'IP {ip_address:IP} is not a internal IP address')
  File "/var/venv-stable/lib/python2.7/site-packages/behave/step_registry.py", line 92, in wrapper
    self.add_step_definition(step_type, step_text, func)
  File "/var/venv-stable/lib/python2.7/site-packages/behave/step_registry.py", line 59, in add_step_definition
    step_definitions.append(get_matcher(func, step_text))
  File "/var/venv-stable/lib/python2.7/site-packages/behave/matchers.py", line 422, in get_matcher
    return current_matcher(func, pattern)
  File "/var/venv-stable/lib/python2.7/site-packages/behave/matchers.py", line 231, in __init__
    self.parser = self.parser_class(pattern, self.custom_types)
  File "/var/venv-stable/lib/python2.7/site-packages/parse.py", line 644, in __init__
    self._expression = self._generate_expression()
  File "/var/venv-stable/lib/python2.7/site-packages/parse.py", line 806, in _generate_expression
    e.append(self._handle_field(part))
  File "/var/venv-stable/lib/python2.7/site-packages/parse.py", line 872, in _handle_field
    format = extract_format(format, self._extra_types)
  File "/var/venv-stable/lib/python2.7/site-packages/parse.py", line 614, in extract_format
    raise ValueError('format spec %r not recognised' % type)
ValueError: format spec u'IP' not recognised
```

**Comment by jenisys** [comment]
Type registration only works when modules are imported.
For example, ensure that type-definition modules are imported in:

*  ``features/environment.py`` or
* step module that needs them before the type is used

**Comment by belugame** [comment]
Thank you for your advice, I have added imports in both places and made sure a breakpoint before the register_type is reached, but it is still failing.

**Comment by belugame** [comment]
I added imports everywhere and between now.
And also a line `print(extra_types.keys())` to the Parser class of the parse pip package.
It prints all my data types from 2 different files nicely at first but at some point the list becomes empty. I have no idea why. :-/

**Comment by jenisys** [comment]
I am pretty sure that the problem is on your side, because [behave.example](https://github.com/behave/behave.example) basically tests this functionality (in addition to behave) with behave-1.2.6. I am also not aware that something has changed, except that a new ``parse`` and ``parse_type`` version is used with behave-1.2.6.

Therefore, isolate everything. 
Make a small example, isolated from your project that either works or does not work.
If it works get closer to your use case until you be able to identify the problem location.

SEE ALSO:
* [behave.example/datatype.features/](https://github.com/behave/behave.example/tree/master/datatype.features)
* [behave.example tutorial 10](http://behave.github.io/behave.example/tutorials/tutorial10.html): [features/steps/step_tutorial10.py](https://github.com/behave/behave.example/blob/master/features/steps/step_tutorial10.py)

**Comment by belugame** [comment]
Thank you very much for your advice, I will try to do that and reopen the issue if needed.

**Comment by belugame** [comment]
Just an update as it may be helpful for others:

The problem was fixed by doing things the behave-intended way. The legacy solution I worked on had features/steps directories per-app in a big django project. As I learned the hard way that just does not work properly with behave, even though it may be neater / closer to the django philosophy.
Now I have a single feature folder and a single steps folder and a single environment.py and the types get found fine.
